yxi/xml/xml_generator.py

```diff
@@ -0,0 +1,113 @@
+import sys
+import json
+from xml.etree.ElementTree import Element, SubElement, tostring
+from xml.dom import minidom
+from math import log2
+
+"""
+This file takes in a `.yxi` description and outputs a xml suitable for a `kernel.xml` file
+can be used to package an xclbin for the Xilinx XRT runtime.
+See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+for the spec this is based on.
+"""
+size_key = "total_size"
+width_key = "data_width"
+
+def gen_xml(yxi):
+  mems = yxi["memories"]
+  check_mems_wellformed(mems)
+
+  root = Element("root", {"versionMajor": "1", "versionMinor": "6"})
+  kernel = SubElement(root, "kernel", {
+    "name" : yxi["toplevel"],
+    "language": "ip_c",
+    # TODO: Make sure this matches component.xml, Namely the `Toplevel` part.
+    #  See https://docs.amd.com/r/en-US/ug1393-vitis-application-acceleration/RTL-Kernel-XML-File
+    "vlnv": "capra.cs.cornell.edu:kernel:Toplevel:1.0",
+    "attributes": "",
+    "preferredWorkGroupSizeMultiple" : "0",
+    "workGroupSize" : "1",
+    "hwControlProtocol": "ap_ctrl_hs"
+  })
+
+  # Construct ports
+  ports = SubElement(kernel, "ports")
+  # The subordinates XRT - AXI controller is added outside of the programs memory interface.
+  SubElement(ports, "port", {
+    "name": "S_AXI_CONTROL",
+    "mode": "slave",
+    "range": "0x1000", # XXX: Why is this 0x1000? Taken from the Xilinx examples  Maybe could be dynamically calculated?
+    "dataWidth" : "32",
+    "portType" : "addressable",
+    "base" : "0x0"
+  })
+
+  for mem in mems:
+    SubElement(ports, "port", {
+      "name": f"m_axi_{mem['name']}",
+      "mode": "master",
+      #NOTE(nathaniel): In the Xilinx examples range is usually 0xFFFFFF... but this should be fine for us?
+      "range": f"{hex(size_in_bytes(mem))}",
+      "dataWidth": f"{mem[width_key]}",
```

Comment:
[Non-actionable:] As you've noted elsewhere, the old version hard-codes the width to 512. Great if we can relax this… but just in case something mysterious goes wrong, we can go back to a fixed size for debugging?